uc-crux-llvm/src/UCCrux/LLVM/Constraints.hs

```diff
@@ -0,0 +1,371 @@
+          if Vec.length vec >= n
+            then (Just InitializeInitialized, shape) -- There's enough space already
+            else -- Grow the allocation
+
+              ( Nothing,
+                Shape.ShapePtr
+                  tag
+                  ( Shape.ShapeInitialized
+                      (vec <> minimalVec (n - Vec.length vec) (asFullType mts ptRepr))
+                  )
+              )
+
+--------------------------------------------------------------------------------
+
+-- * 'NewConstraint'
+
+data ExpansionError
+  = ExpRedundant RedundantExpansion
+  | ExpShapeSeekError ShapeSeekError
+
+ppExpansionError :: ExpansionError -> Text
+ppExpansionError =
+  \case
+    ExpRedundant redundant -> ppRedundantExpansion redundant
+    ExpShapeSeekError ssErr -> Shape.ppShapeSeekError ssErr
+
+-- | A constraint learned by simulating the code
+data NewConstraint m argTypes
+  = forall atTy. NewShapeConstraint (SomeInSelector m argTypes atTy) (ShapeConstraint m atTy)
+  | forall atTy. NewConstraint (SomeInSelector m argTypes atTy) (Constraint m atTy)
+  | NewRelationalConstraint (RelationalConstraint m argTypes)
+
+addConstraint ::
+  Ctx.Assignment (FullTypeRepr m) argTypes ->
+  ModuleTypes m ->
+  Constraints m argTypes ->
+  NewConstraint m argTypes ->
+  Either ExpansionError (Constraints m argTypes)
+addConstraint argTypes mts constraints =
+  \case
+    NewConstraint (SomeInSelector (SelectGlobal _symbol _cursor)) _constraint ->
+      unimplemented "addConstraint" Unimplemented.ConstrainGlobal
+    NewShapeConstraint (SomeInSelector (SelectGlobal _symbol _cursor)) _shapeConstraint ->
+      unimplemented "addConstraint" Unimplemented.ConstrainGlobal
+    NewConstraint (SomeInSelector (SelectArgument idx cursor)) constraint ->
+      constraints & argConstraints . ixF' idx
+        %%~ ( \(ConstrainedShape shape) ->
+                ConstrainedShape
+                  <$> mapLeft
+                    ExpShapeSeekError
+                    (fst <$> Shape.modifyTag (coerce (constraint :)) shape cursor)
+            )
+    NewShapeConstraint (SomeInSelector (SelectArgument idx cursor)) shapeConstraint ->
+      constraints & argConstraints . ixF' idx
+        %%~ ( \(ConstrainedShape shape) ->
+                ConstrainedShape
+                  <$> joinLeft
+                    ExpShapeSeekError
+                    ( Shape.modifyA'
+                        ( \shape' ->
+                            case expand mts (Compose []) (seekType mts cursor (argTypes ^. ixF' idx)) shapeConstraint shape' of
+                              -- TODO(lb): Maybe warn here, there's a risk of an
+                              -- infinite loop if there's a bug in
+                              -- classification.
+                              (Just _redundant, result) -> Right result
+                              (Nothing, result) -> Right result
+                        )
```

Comment:
I have a few other places I'd like to ask this, but can we pull out and name these big nested cases? It would make the arguments to/structure of `modifyA'` easier to eyeball